### ISSUE TITLE
ModelCheckpoint reduce logic seems wrong

### ISSUE THREAD

**Issue by manbango** [Lightning-AI/pytorch-lightning#20141]
### Bug description

The reduction logic in ModelCheckpoint seems wrong.

Specifically this line:
```python
should_update_best_and_save = trainer.strategy.reduce_boolean_decision(bool(should_update_best_and_save))
```

The all condition seems too strong here, we only care if the average of the metric across all workers is lower correct?

I see entries in my tensorboard which are clearly lower loss, but due to randomness, some workers might disagree, so a checkpoint is not saved.

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

Train a model for which loss varies in a relatively small range with many workers.

### Error messages and logs

```
# Error messages and logs here please
```
No Error Messages

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 1.5.0): 2.3.3
#- PyTorch Version (e.g., 2.0): 2.4
#- Python version (e.g., 3.9):  3.10
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version: 12
#- GPU models and configuration: A100
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

cc @carmocca @awaelchli

**Comment by awaelchli** [comment]
@manbango This looks correct to me. The model checkpoint callback just ensures that the decision is synchronized here, it has no say in how the metric is interpreted or synced. That's done through the logging system.

If you'd like to have a metric computed as an average over the world size, you can log it with `self.log(..., sync_dist=True)` (naive average) or use [TorchMetrics](https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html). Then the ModelCheckpoint will get that synchronized metric when you set it as monitor, and it'll make it's decisions based on that.

**Comment by manbango** [comment]
But in the case where the metric is not synced,  and is computed based on the local split of the worker. Why would we want to ensure they all return the same decision?

This seem wrong because it requires that the max value across all workers is less than the previous best metric value. It's not clear to me why you'd want that behavior?

**Comment by awaelchli** [comment]
> But in the case where the metric is not synced, and is computed based on the local split of the worker. Why would we want to ensure they all return the same decision?

The alternative of letting any rank decide to checkpoint isn't so great. Imagine you checkpoint based on a metric when it's decreasing. Suppose you have two GPUs (two ranks) and rank 0's metric hasn't improved, but rank 1's metric has improved. The current behavior is not to update the top-k ranking. However, if it did do that like you describe, then the current weights would be marked as top-1 and a checkpoint gets saved. The problem is of course that the metric is logged on rank 0 and thus the user would see the metric increase while at the same time the user sees Lightning saving a top-1 checkpoint. That would certainly be flagged as a bug!

> This seem wrong because it requires that the max value across all workers is less than the previous best metric value. It's not clear to me why you'd want that behavior?

It's not that we want that behavior necessarily, it's just the basic implementation and a choice that was made to keep the state in sync. Ideally, users would use TorchMetrics to compute metrics correctly across ranks, which is what we recommend always.

**Comment by manbango** [comment]
@awaelchli I understand.
One of the following seem like more reasonable choices to me:

- only decide based on rank 0: if the user has synced properly there’s no difference, if the user hasn’t synced properly then at least it will reflect the logs

- check if a majority of decisions pass, not all, this in some sense is the best estimator so it seems like a reasonable choice to me, with no extra overhead

I get the team may not want to change this logic, but it just was counterintuitive to me when I saw it, so just thought it’s worth looking at.

**Comment by awaelchli** [comment]
> only decide based on rank 0: if the user has synced properly there’s no difference, if the user hasn’t synced properly then at least it will reflect the logs

It's a viable alternative yes. It was like this before: For context here is the PR that changed it to the current logic: https://github.com/Lightning-AI/pytorch-lightning/commit/0544efd4535cb8965902fc3c938263b3deafc7b7

> I get the team may not want to change this logic

We only need to be careful not to regress on prior issues and the current expectations. A change like this can be made if it's desired by the community of course.

**Comment by manbango** [comment]
Thanks for the careful consideration and conversation!
